### PR TITLE
PERF: remove use of Python sets for interpolate

### PR DESCRIPTION
**Pull request by simonjayhawkins**
have been investigating avoiding internals for interpolation. xref #34628

This PR address the issue of using Python sets, which is responsible for the bulk of the time in our current asv.

There are other improvements (will raise other PRs), but would need new benchmarks to show a benefit such as different index types and unsorted indexes. so this PR is targeting our current benchmark first.

prelim results (will post asv results if tests pass)
```
N = 10000
# this is the worst case, where every column has NaNs.
df = pd.DataFrame(np.random.randn(N, 100))
df.values[::2] = np.nan
df
%timeit df.interpolate()
# 189 ms ± 24.1 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)  <-- master
# 65.1 ms ± 635 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)  <-- PR
```

### CODE REVIEW

**Comment by simonjayhawkins** [comment]
some environments are failing with `TypeError: diff() got an unexpected keyword argument 'prepend'`

File: pandas/core/missing.py [hunk 1]
```diff
@@ -678,10 +660,15 @@ def clean_reindex_fill_method(method):
     return clean_fill_method(method, allow_nearest=True)
 
 
-def _interp_limit(invalid, fw_limit, bw_limit):
+def _interp_limit(
+    invalid: np.ndarray,
+    fw_limit: Optional[int],
+    bw_limit: Optional[int],
+    first: int,
```
Comment (simonjayhawkins) [inline pandas/core/missing.py:667]: i'll update the algorithm so that passing first and last are unnecessary.

File: pandas/core/missing.py [hunk 2]
```diff
@@ -225,41 +225,25 @@ def interpolate_1d(
...
     if limit_direction == "forward":
-        preserve_nans = start_nans | set(_interp_limit(invalid, limit, 0))
+        nans_to_interpolate = _interp_limit(invalid, limit, 0, first, last)
     elif limit_direction == "backward":
-        preserve_nans = end_nans | set(_interp_limit(invalid, 0, limit))
+        nans_to_interpolate = _interp_limit(invalid, 0, limit, first, last)
     else:
         # both directions... just use _interp_limit
-        preserve_nans = set(_interp_limit(invalid, limit, limit))
+        nans_to_interpolate = _interp_limit(invalid, limit, limit, first, last)
```
Comment (simonjayhawkins) [inline pandas/core/missing.py:236]: we could move the above logic into _interp_limit and fastpath if limit_area.

although the algorithm could probably be extended to 2d, in which case the limit_area logic below would also need to move.

File: pandas/core/missing.py (hunk 1 above)
Comment (simonjayhawkins) [reply pandas/core/missing.py]: https://github.com/pandas-dev/pandas/pull/34727/commits/1702ef50b63086b01e5b7aa4254d631d71b91113

File: pandas/core/missing.py [hunk 3]
```diff
@@ -681,68 +666,121 @@ def _interp_limit(invalid, fw_limit, bw_limit):
...
+    >>> arr.astype(bool)
+    array([ False,  False, False, False,  True,  True, False, False,  True,
+            True])
     """
-    # https://stackoverflow.com/a/6811241
-    shape = a.shape[:-1] + (a.shape[-1] - window + 1, window)
-    strides = a.strides + (a.strides[-1],)
-    return np.lib.stride_tricks.as_strided(a, shape=shape, strides=strides)
+
+    def inner(arr, limit):
```
Comment (simonjayhawkins) [inline pandas/core/missing.py:758]: will move this into a top level function called ffill_mask_with_limit (in this PR)

and plan to take the max_gap logic from #25141 and have a analogous ffill_mask_with_max_gap (separate PR)

File: pandas/core/missing.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/core/missing.py]: > although the algorithm could probably be extended to 2d, in which case the limit_area logic below would also need to move.

i don't think a 2d version will be needed as interpolate_2d will also be applied along axis when max_gap is added. 

> we could move the above logic into _interp_limit and fastpath if limit_area.

probably best as a follow-on to keep the diff here smaller

**Comment by jreback** [comment]
looks fine.

**Comment by jreback** [comment]
do we have sufficient asv's for this?

**Comment by simonjayhawkins** [comment]
will get back to this after 1.0.5 we don't have many asvs so could probably add more The current asv is a single float block with a sorted RangeIndex and no limit (which can be fastpathed) so not representative of potentially much slower cases

**Comment by simonjayhawkins** [comment]
I'll close this for now. still don't have access to desktop and running asvs on laptop is painful.

**Pull request closed without merging**